srcts/src/shiny/bind.ts

```diff
@@ -47,45 +48,54 @@ function valueChangeCallback(
  * immediately invoked function to keep the sets private and not clutter the
  * scope.
  */
-const bindingsRegistery = (() => {
+const bindingsRegistry = (() => {
+  type BindingTypes = "input" | "output";
+
   /**
    * Keyed by binding IDs to the array of each type of binding that ID is associated for in current app state.
    *
    * Ideally the
    * value would be a length 1 array but in some (invalid) cases there could be
    * multiple types for a single ID.
    */
-  type IdToBindingTypes = Map<string, Array<"input" | "output">>;
+  type IdToBindingTypes = Map<string, BindingTypes[]>;
 
   // Main store of bindings.
   const bindings: IdToBindingTypes = new Map();
 
   /**
-   * Checks if the bindings registery is valid. Currently this just checks for
+   * Checks if the bindings registry is valid. Currently this just checks for
    * duplicate IDs but in the future could be expanded to check more conditions
-   * @returns ShinyClientError if current ID bindings are invalid, otherwise null
+   * @returns ShinyClientError if current ID bindings are invalid, otherwise
+   * returns an ok status.
    */
   function checkValidity():
     | { status: "error"; error: ShinyClientError }
     | { status: "ok" } {
-    const duplicateIds: IdToBindingTypes = new Map();
+    const duplicateIds: Map<string, { input: number; output: number }> =
+      new Map();
+
+    const countBindingType = (
+      types: BindingTypes[],
+      bindingType: BindingTypes
+    ) => {
+      return types.filter((type) => type === bindingType).length;
+    };
 
-    bindings.forEach((inputOrOutput, id) => {
-      if (inputOrOutput.length > 1) {
-        duplicateIds.set(id, inputOrOutput);
+    // count duplicate IDs of each binding type
+    bindings.forEach((idTypes, id) => {
+      const nInputs = countBindingType(idTypes, "input");
+      const nOutputs = countBindingType(idTypes, "output");
+
+      if (nInputs > 1 || nOutputs > 1) {
+        duplicateIds.set(id, { input: nInputs, output: nOutputs });
       }
     });
```

Comment:
```suggestion
    // count duplicate IDs of each binding type
    bindings.forEach((idTypes, id) => {
      const counts: { [T in BindingTypes]: number } = { input: 0, output: 0 };

      idTypes.forEach((type) => (counts[type] += 1));

      if (counts.input > 1 || counts.output > 1) {
        duplicateIds.set(id, counts);
      }
    });
```

Reply:
Just a couple of suggestions for avoiding an extra abstraction and also keeping nice type checking with the new `BindingTypes` type.

Reply:
I'm much more attached to the comment explaining the somewhat perplexing-out-of-context `if` statement than anything here though.

Reply:
I rewrote the logic to check if any count is `> 1`, not just input and output counts. I think the code itself is clear enough, but I agree we need to record the relevant parts of this discussion. I added that into the jsdoc so that it's close to the code and also shows up in the hover cards. (688334402d68e5f88e4d6f980cacc9b03da8dc75)

And to clarify, I think we _do_ want to allow duplicate IDs across inputs and outputs; what we don't want to do is to have more than one element on the page with the same `id`. In other words: users shouldn't have to think of inputs and outputs as a shared namespace. I think the onus is us as the framework authors to make the markup work.